Mathlib/Algebra/BigOperators/Finprod.lean

```diff
@@ -0,0 +1,1292 @@
+* `∏ᶠ (n >= -2) (hn : n < 3), f n`, where `f : ℤ → M` : same as `f (-2) * f (-1) * f 0 * f 1 * f 2`.
+
+## Implementation notes
+
+`Finsum` and `Finprod` is "yet another way of doing finite sums and products in Lean". However
+experiments in the wild (e.g. with matroids) indicate that it is a helpful approach in settings
+where the user is not interested in computability and wants to do reasoning without running into
+typeclass diamonds caused by the constructive finiteness used in definitions such as `Finset` and
+`Fintype`. By sticking solely to `Set.finite` we avoid these problems. We are aware that there are
+other solutions but for beginner mathematicians this approach is easier in practice.
+
+Another application is the construction of a partition of unity from a collection of “bump”
+function. In this case the finite set depends on the point and it's convenient to have a definition
+that does not mention the set explicitly.
+
+The first arguments in all definitions and lemmas is the codomain of the function of the big
+operator. This is necessary for the heuristic in `@[to_additive]`.
+See the documentation of `to_additive.attr` for more information.
+
+We did not add `isFinite (X : Type) : Prop`, because it is simply `Nonempty (Fintype X)`.
+
+## Tags
+
+finsum, finprod, finite sum, finite product
+-/
+
+
+open Function Set
+
+/-!
+### Definition and relation to `Finset.sum` and `Finset.prod`
+-/
+
+-- Porting note: Used to be section Sort
+section sort
+
+variable {G M N : Type _} {α β ι : Sort _} [CommMonoid M] [CommMonoid N]
+
+-- Porting note: Unknown namespace BigOperators
+--open BigOperators
+
+section
+
+/- Note: we use classical logic only for these definitions, to ensure that we do not write lemmas
+with `Classical.dec` in their statement. -/
+open Classical
+
+
+-- Porting note: replaced irreducible_def with def and an irreducible tag here.
+/-- Sum of `f x` as `x` ranges over the elements of the support of `f`, if it's finite. Zero
+otherwise. -/
+@[irreducible]
+noncomputable def finsum {M α} [AddCommMonoid M] (f : α → M) : M :=
+  if h : (support (f ∘ PLift.down)).Finite then ∑ i in h.toFinset, f i.down else 0
+#align finsum finsum
+
+-- Porting note: replaced irreducible_def with def and an irreducible tag here.
+/-- Product of `f x` as `x` ranges over the elements of the multiplicative support of `f`, if it's
+finite. One otherwise. -/
+@[to_additive (attr:= irreducible)]
+noncomputable def finprod (f : α → M) : M :=
+  if h : (mulSupport (f ∘ PLift.down)).Finite then ∏ i in h.toFinset, f i.down else 1
+#align finprod finprod
+
+end
+
+-- mathport name: finsum
+-- scoped[BigOperators] -- Porting note: `notation3` doesn't mesh with `scoped[Foo]`
+@[inherit_doc finsum]
+notation3"∑ᶠ "(...)", "r:(scoped f => finsum f) => r
+
+-- mathport name: finprod
+-- scoped[BigOperators] -- Porting note: `notation3` doesn't mesh with `scoped[Foo]`
+@[inherit_doc finprod]
+notation3"∏ᶠ "(...)", "r:(scoped f => finprod f) => r
```

Comment:
Please take a look at how the notation was set up for the `finset.{prod,sum}`. That will also allow you to set operator precedence, which is probably important.

Reply:
I had a go at fixing the notation. It's not that easy, I say we merge as is.

Reply:
I added some macro_rules  in comments that would fix the notation in this file, and in this file only; I do not know the necessary tools to produce automatic macro_rules that would allow the product over arbitrarily many arguments, or to ignore optional type hints. If someone does, I reckon they can modify the comments to fix it, but at least functionally nothing changes.